### PR TITLE
Pairwise contour plots of GP fit with more than 2 parameters

### PR DESCRIPTION
**Pull request by b5y**
#### Summary:
Add feature to run GP model with more than 2 parameters

Example of usage: https://github.com/elfi-dev/zoo/blob/gauss_plots/vis/bolfi_gpy.ipynb

### CODE REVIEW

File: elfi/methods/parameter_inference.py [hunk 1]
```diff
@@ -1132,6 +1132,20 @@ def plot_discrepancy(self, axes=None, **kwargs):
...
+        ----------
+        axes : matplotlib.axes.Axes, optional
+
+        Return
+        ------
+        axes : np.array of plt.Axes
+
+        """
+        return vis.plot_gp(self.target_model, self.parameter_names, axes, **kwargs)
+
```
Comment (vuolleko) [inline elfi/methods/parameter_inference.py]: It would be good to move all plotting out from this file, for example by using mixin classes.

Can be a separate PR though.

File: CHANGELOG.rst [hunk 2]
```diff
@@ -1,6 +1,10 @@
 Changelog
 =========
 
+dev
+---
+- Implement plot_gp for BOLFI for plotting parameters vs. contours for more than 2 parameters
```
Comment (vuolleko) [inline CHANGELOG.rst]: I think this should work for the BayesianOptimization class as well.

File: elfi/visualization/visualization.py [hunk 3]
```diff
@@ -359,3 +359,50 @@ def plot_params_vs_node(node, n_samples=100, func=None, seed=None, axes=None, **
...
+        Parameter names from model.parameters dict('parameter_name':(lower, upper), ... )`
+    axes : matplotlib.axes.Axes, optional
+
+    Return
+    ------
+    axes : np.array of plt.Axes
+
+    """
+    n_plots = gp.input_dim
+    resol = 50
```
Comment (vuolleko) [inline elfi/visualization/visualization.py]: Should be user-given (50 is fine as default).

File: elfi/visualization/visualization.py [hunk 4]
```diff
@@ -359,3 +359,50 @@ def plot_params_vs_node(node, n_samples=100, func=None, seed=None, axes=None, **
...
+    shape = (n_plots, n_plots)
+    axes, kwargs = _create_axes(axes, shape, **kwargs)
+
+    x_evidence = np.random.rand(10, n_plots) + np.arange(n_plots)
+    y_evidence = gp.predict_mean(x_evidence)
+    const = x_evidence[np.argmin(y_evidence), :]
```
Comment (vuolleko) [inline elfi/visualization/visualization.py]: Should be user-given (ok as default).

File: elfi/visualization/visualization.py [hunk 5]
```diff
@@ -397,3 +397,50 @@ def plot_discrepancy(gp, parameter_names, axes=None, **kwargs):
...
+
+def plot_gp(gp, parameter_names, axes=None, resol=50, const=None, **kwargs):
+    """Plot pairwise relationships as a matrix with parameters vs. discrepancy.
+
+    Parameters
+    ----------
+    gp : GPyRegression, required
+    parameter_names : dict, required
+        Parameter names from model.parameters dict('parameter_name':(lower, upper), ... )`
+    axes : matplotlib.axes.Axes, optional
```
Comment (vuolleko) [inline elfi/visualization/visualization.py]: Single axis or an array? If single, how will it be used?

File: elfi/visualization/visualization.py [hunk 6]
```diff
@@ -397,3 +397,50 @@ def plot_discrepancy(gp, parameter_names, axes=None, **kwargs):
...
+
+
+def plot_gp(gp, parameter_names, axes=None, resol=50, const=None, **kwargs):
+    """Plot pairwise relationships as a matrix with parameters vs. discrepancy.
+
+    Parameters
+    ----------
+    gp : GPyRegression, required
+    parameter_names : dict, required
+        Parameter names from model.parameters dict('parameter_name':(lower, upper), ... )`
```
Comment (vuolleko) [inline elfi/visualization/visualization.py]: Names or bounds? If just the names, why is a dict required?

File: elfi/methods/parameter_inference.py [hunk 7]
```diff
@@ -1126,6 +1126,24 @@ def plot_discrepancy(self, axes=None, **kwargs):
...
+            Used to construct np.meshgrid and arrays of predictors for contours.
+        const : np.array, optional
+            Used to construct array of predictors for contours.
+
+        Returns
+        -------
+        axes : np.array of plt.Axes
+
+        """
+        return vis.plot_gp(self.target_model, self.parameter_names, axes, resol, const, **kwargs)
```
Comment (vuolleko) [inline elfi/methods/parameter_inference.py]: The signature does not match (self.parameter_names is a list).

File: elfi/visualization/visualization.py [hunk 8]
```diff
@@ -397,3 +397,50 @@ def plot_discrepancy(gp, parameter_names, axes=None, **kwargs):
...
+    n_plots = gp.input_dim
+    shape = (n_plots, n_plots)
+    axes, kwargs = _create_axes(axes, shape, **kwargs)
+
+    x_evidence = gp.X
+    y_evidence = gp.Y
+    if not const:
+        const = x_evidence[np.argmin(y_evidence), :]
+    bounds_min = x_evidence.min(axis=0)
+    bounds_max = x_evidence.max(axis=0)
```
Comment (vuolleko) [inline elfi/visualization/visualization.py]: Could use the bounds given by the user, if the description is signature is correct. The `gp` object also has a `bounds` attribute.

File: CHANGELOG.rst [hunk 9]
```diff
@@ -4,6 +4,8 @@ Changelog
 dev
 ---
 - Fix bug in plot_discrepancy for more than 6 parameters
+- Implement plot_gp for BayesianOptimization classes for plotting parameters vs. contours
+  in case when we have more than 2 parameters
```
Comment (vuolleko) [inline CHANGELOG.rst]: "parameters vs. contours" -> "discrepancies and pair-wise contours"

Should work with any number of parameters?

File: elfi/visualization/visualization.py (hunk 6 above)
Comment (hpesonen) [reply elfi/visualization/visualization.py]: I think parameter_names should list or tuple instead of dict

File: elfi/visualization/visualization.py (hunk 8 above)
Comment (b5y) [reply elfi/visualization/visualization.py]: Which one is better in this case? `gp.bounds` or giving as a parameter from user?

File: elfi/visualization/visualization.py (hunk 6 above)
Comment (b5y) [reply elfi/visualization/visualization.py]: It is a _list_.

File: elfi/visualization/visualization.py (hunk 5 above)
Comment (b5y) [reply elfi/visualization/visualization.py]: `plt.Axes or arraylike of plt.Axes`. The usage is the same as in other functions in _visualization_ module.

File: elfi/visualization/visualization.py [hunk 10]
```diff
@@ -397,3 +397,50 @@ def plot_discrepancy(gp, parameter_names, axes=None, **kwargs):
...
+    axes : np.array of plt.Axes
+
+    """
+    n_plots = gp.input_dim
+    shape = (n_plots, n_plots)
+    axes, kwargs = _create_axes(axes, shape, **kwargs)
+
+    x_evidence = gp.X
+    y_evidence = gp.Y
+    if not const:
```
[the remark this answers is not in the document]
Comment (b5y) [reply elfi/visualization/visualization.py]: I didn't know that we consider the case when `const = 0`.

[the remark this answers is not in the document]
Comment (vuolleko) [reply elfi/visualization/visualization.py]: Well `const` can be anything the user wants. :)

File: elfi/visualization/visualization.py (hunk 8 above)
Comment (vuolleko) [reply elfi/visualization/visualization.py]: Default to `gp.bounds` but allow the user to override.

File: CHANGELOG.rst (hunk 9 above)
Comment (vuolleko) [reply CHANGELOG.rst]: I meant that this should work for cases of 1 and 2 parameters also. :D

Comment (b5y) [reply CHANGELOG.rst]: Well, I always check plottings from 1 up to 8-9 parameters if there is no discussion with parameters :)

Comment (b5y) [reply CHANGELOG.rst]: The reason why I wrote parameters number is because we had this problem. I will fix it with 'arbitrary' number of parameters (but not larger than 8).

File: elfi/visualization/visualization.py (hunk 8 above)
Comment (b5y) [reply elfi/visualization/visualization.py]: `gp.bounds` is defined by user. It is a list of tuples which defines intervals on grid. What is the benefit of using `gp.bounds` instead of `x_evidence` which is taken from `gp.X`? I understand that we should give more freedom for users in case of defining models, but changing with `gp.bounds` is not clear for me.

File: elfi/visualization/visualization.py [hunk 11]
```diff
@@ -397,3 +397,49 @@ def plot_discrepancy(gp, parameter_names, axes=None, **kwargs):
...
+    """
+    n_plots = gp.input_dim
+    shape = (n_plots, n_plots)
+    axes, kwargs = _create_axes(axes, shape, **kwargs)
+
+    x_evidence = gp.X
+    y_evidence = gp.Y
+    if not const and const != 0:
+        const = x_evidence[np.argmin(y_evidence), :]
+    bounds = kwargs.get('bounds', None) or gp.bounds
```
[the remark this answers is not in the document]
Comment (b5y) [reply elfi/visualization/visualization.py]: It makes sense to create new function if and only if the object is used more than 2 times. 
Okay, I will add some documentation above this line.

File: elfi/visualization/visualization.py [hunk 12]
```diff
@@ -397,3 +397,49 @@ def plot_discrepancy(gp, parameter_names, axes=None, **kwargs):
...
+
+    for ix in range(n_plots):
+        for jy in range(n_plots):
+            if ix == jy:
+                axes[jy, ix].scatter(x_evidence[:, ix], y_evidence)
+                axes[jy, ix].set_xlim(bounds[ix])
+                axes[jy, ix].set_ylabel('Discrepancy')
+            else:
+                x1 = np.linspace(min(bounds[ix]), max(bounds[ix]), resol)
```
[the remark this answers is not in the document]
Comment (b5y) [reply elfi/visualization/visualization.py]: Yes, in first case I was thinking to make it like this.
In my opinion, it is more evident and readable to keep `min` and `max` functions because it allows not to go into `linspace` definition and read what it is doing.

File: elfi/visualization/visualization.py [hunk 13]
```diff
@@ -397,3 +397,49 @@ def plot_discrepancy(gp, parameter_names, axes=None, **kwargs):
         axes[idx].set_axis_off()
 
     return axes
+
+
+def plot_gp(gp, parameter_names, axes=None, resol=50, const=None, **kwargs):
+    """Plot pairwise relationships as a matrix with parameters vs. discrepancy.
+
+    Parameters
```
[the remark this answers is not in the document]
Comment (b5y) [reply elfi/visualization/visualization.py]: Well, I think it is time to add some [docstring checking](https://github.com/PyCQA/pydocstyle) in Makefile. Otherwise we should skip double documentation in `defining` and `calling` functions.

File: CHANGELOG.rst [hunk 14]
```diff
@@ -4,6 +4,8 @@ Changelog
 dev
 ---
 - Fix bug in plot_discrepancy for more than 6 parameters
+- Implement plot_gp for BayesianOptimization classes for plotting discrepancies
+  and pair-wise contours in case when we have arbitrary number of parameters (from 1 up to 8)
```
[the remark this answers is not in the document]
Comment (b5y) [reply CHANGELOG.rst]: In aforementioned given notebook file I checked this code up to 8 parameters. As far as I remember, we argued that it is enough. If we don't this in change-log file, I will remove it.

File: elfi/visualization/visualization.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (b5y) [reply elfi/visualization/visualization.py]: To be honest, I have never seen in any `function` or `class` (**in ELFI**) anything to be documented about keyword arguments explicitly. Please give me a link if I am wrong.

File: CHANGELOG.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (vuolleko) [reply CHANGELOG.rst]: Yes, just remove the part "(from 1 up to 8)".

File: elfi/visualization/visualization.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (vuolleko) [reply elfi/visualization/visualization.py]: Inheriting docstrings would be great. But until then, explicit is better than implicit. :)

File: elfi/visualization/visualization.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (vuolleko) [reply elfi/visualization/visualization.py]: What new function? Why do you say keyword arguments are never documented?

I just mean add `bounds` in the signature.

File: elfi/visualization/visualization.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (vuolleko) [reply elfi/visualization/visualization.py]: Redundant, but ok.
